mkosi/run.py

```diff
@@ -209,21 +213,56 @@ def spawn(
     if "HOME" not in env:
         env["HOME"] = "/"
 
-    with sandbox as sbx:
-        prefix = [os.fspath(x) for x in sbx]
+    with contextlib.ExitStack() as stack:
+        sbx = [os.fspath(x) for x in stack.enter_context(sandbox)]
+        apply_sandbox_in_preexec = not setup and not ARG_DEBUG_SANDBOX.get()
+
+        def _preexec() -> None:
+            if preexec:
+                preexec()
+
+            if sbx and apply_sandbox_in_preexec:
+                # The env passed to subprocess.Popen() replaces the environment wholesale so any
+                # modifications made by mkosi-sandbox would be overridden if we used that. Hence process the
+                # environment in the preexec function.
+                os.environ.clear()
+                os.environ.update(env)
+                try:
+                    mkosi.sandbox.main(sbx)
+                except Exception:
+                    sys.excepthook(*ensure_exc_info())
+                    os._exit(1)
+
+        prefix: list[str]
+        if sbx and not apply_sandbox_in_preexec:
+            module = stack.enter_context(resource_path(sys.modules[__package__ or __name__]))
+            prefix = [
+                *(["strace", "--detach-on=execve"] if ARG_DEBUG_SANDBOX.get() else []),
+                sys.executable, "-SI", os.fspath(module / "sandbox.py"),
+                *sbx,
+            ]  # fmt: skip
+        else:
+            prefix = []
```

Comment:
Since the typing needs to stay outside anyway, we can use this to get rid of the else branch and make this a bit easier on the eyes. 
```suggestion
        prefix: list[str] = []
        if sbx and not apply_sandbox_in_preexec:
            module = stack.enter_context(resource_path(sys.modules[__package__ or __name__]))
            prefix = [
                *(["strace", "--detach-on=execve"] if ARG_DEBUG_SANDBOX.get() else []),
                sys.executable, "-SI", os.fspath(module / "sandbox.py"),
                *sbx,
            ]  # fmt: skip
```